### ISSUE TITLE
How to handle pong messages? Bug?

### ISSUE THREAD

**Issue by tjeske** [gorilla/websocket#667]
**Describe the problem you're having**
I want to implement a health check for my server to detect clients which are gone. I send a ws ping message to the client (browser) which responds with a ws pong message. If the pong message does not arrive I want to free some internal server resources.
The problem is that I always run into a timeout in `NextReader()` after I have called `WriteMessage(websocket.PingMessage, []byte{})`.

**Versions**
Go version: `1.16`
package version: `1.4.2`

**"Show me the code!"**

```
c.WriteMessage(websocket.PingMessage, []byte{})
..
c.SetReadLimit(8192)
c.SetReadDeadline(time.Now().Add(20 * time.Second))
c.SetPongHandler(func(string) error {
    // this is called only once
    c.SetReadDeadline(time.Now().Add(20 * time.Second))
    return nil
})
_, message, err := c.NextReader() // this blocks for 20s and then returns with an errror
```

**Comment by MonicaRJones** [comment]
The application is working as expected.  The application sets the read deadline to 20 seconds in the future and resets the deadline to 20 minutes in the future on each pong.  The connection will timeout if the peer does not send a message every 20 seconds.

Is the application sending pongs more frequently than every 20 seconds?

**Comment by tjeske** [comment]
My server sends pings every two seconds (just for testing) and the browser always responses with pong messages. I have verified it with Wireshark.
I had a look into the code. The pong handler can only be called by `advanceFrame()` which is called by `NextReader()` or `Read()`. The only non-error way, of leaving the for-loop in `NextReader()` is if `frameType == TextMessage || frameType == BinaryMessage` which cannot be the case here because we are receiving pong messages. This explains why it stucks for  20s and then only errors appear. Is calling `NextReader()` the right way to handle pong messages? Is my understanding on how to use the library wrong or is this a bug?

**Comment by MonicaRJones** [comment]
Yes, calling NextReader is the right way to process control messages including PONG messages.  The [relevant section of the documentation(https://pkg.go.dev/github.com/gorilla/websocket#hdr-Control_Messages) says:

> The control message handler functions are called from the NextReader, ReadMessage and message reader Read methods.

and

> The application must read the connection to process close, ping and pong messages sent from the peer.

The sequence of statements 

```
c.SetReadDeadline(time.Now().Add(20 * time.Second))
c.SetPongHandler(func(string) error {
    // this is called only once
    c.SetReadDeadline(time.Now().Add(20 * time.Second))
    return nil
})
_, message, err := c.NextReader()
if err != nil {
     // clean up after client
}
```

is the correct way to detect a peer that's unresponsive for 20 seconds.  There's no need to distinguish timeout from other errors because the application should do the same thing in all errors: clean up resources used by the client and close the connection.

**Comment by tjeske** [comment]
The browser (peer) is responsive (it sends pong messages). The application sends pings, browser responses immediately with pongs. This is all correct but I don't understand why it is expected behaviour that the application run into an timeout (the pongs are not coming in after 20s). It seams to be that the application expects a data message but if this is the case I don't see the point of having ping/pong messages at all.

**Comment by MonicaRJones** [comment]
> the pongs are not coming in after 20s

This sounds like the problem.  Show the code that sends the PING messages.

**Comment by tjeske** [comment]
It is in the code example above (`c.WriteMessage(websocket.PingMessage, []byte{})`).

**Comment by MonicaRJones** [comment]
I misunderstood you wen you said the server sends a ping message every two seconds for testing.   I thought the "for testing" was in references to the extremely short ping period, not that the application does not send pings to keep the reader from timing out.

The application must send pings more frequently than every 20 seconds.

**Comment by tjeske** [comment]
I think this is the misunderstanding. The reason for sending pings in gorilla is to extend the read deadline. The application must expect some data from the peer. Im my scenario this is not the case. I just want to find out if the peer is alive without any data messages involved.

**Comment by MonicaRJones** [comment]
> the application must expect some data from the peer. Im my scenario this is not the case.

The [documentation says](https://pkg.go.dev/github.com/gorilla/websocket#hdr-Control_Messages):

> The application must read the connection to process close, ping and pong messages sent from the peer. If the application is not otherwise interested in messages from the peer, then the application should start a goroutine to read and discard messages from the peer.

Here's an example with a 20 second timeout:

 ```
func readLoop(c *websocket.Conn) {
    defer c.Close()
    c.SetPongHandler(func(string) error {
        c.SetReadDeadline(time.Now().Add(20 * time.Second))
        return nil
    })

    for {
        c.SetReadDeadline(time.Now().Add(20 * time.Second))
        if _, _, err := c.NextReader(); err != nil {
            c.Close()
            return
        }
    }
}
```

**Comment by tjeske** [comment]
Your code expects that the peer is sending something other than a pong message which is **not** the case in my scenario.

**Comment by MonicaRJones** [comment]
The function does not assume that the peer will send a data message.  If the peer does send a data message, then the message is discarded.

If the peer does not send a PONG or data message every 20 seconds, then function closes the connection.    

Insert code into the deferred call to notify other parts of the application that the client is dead.  Here's an example that uses context cancelation:

```
func readLoop(c *websocket.Conn, cancel context.CancelFunc) {
   defer func() {
          c.Close()
          cancel()
   }()
   c.SetPongHandler(func(string) error {
       c.SetReadDeadline(time.Now().Add(20 * time.Second))
       return nil
   })

   for {
       c.SetReadDeadline(time.Now().Add(20 * time.Second))
       if _, _, err := c.NextReader(); err != nil {
           return
       }
   }
}

```

**Comment by tjeske** [comment]
Ah okay, I think I got it. I have written a small test program. It is expected that the application constantly sending pings to the peer (constantly means < read deadline, here 20s). I always thought about a way to probe all peers (e.g. every 20s) and kick every client which is not responding in e.g. 2s.
Thanks for helping me. I appreciate this.
